(basics-calculator)=

# Python as a calculator

The first thing we learn is just that you can use a programming language like a pocket calculator.

This is more important than you think, as there are lots of multiplications throughout all of ecological economics.



### 📖 Basic operations overview

In brief, python has all the standard operations that you know since primary school

:::{admonition} Basic operations

- addition +
- subtraction -
- divison /
- multiplication \*
- raising a number to an exponent \*\*
- how often does a number fit into another number but not considering remainder //
- remaining part of the division %
  :::

you can multiply numbers like in the following code box.

In [1]:
2 + 2, 10 - 3, 6 * 7, 22 / 7, 22 // 7, 22 % 7, 2 ** 10

(4, 7, 42, 3.142857142857143, 3, 1, 1024)

What is shown below in the brackets is the output for each calculation separated by commata.


The ** is raising a number to a power. So 2\*2 = 4 could also be 2**2 because 2 squared is 4.

The % operator returns the remainder of a division. For example, 7 % 2 = 1 because dividing 7 by 2 gives 3 with a remainder of 1 (since
2\*3+1=7).

Got it? With this one you need to become really familiar.

Instead of boring stuff like this, let us do some interesting thought experiments to get into the groove:

The US has a approximately a total primary energy consumption of 100 Exajoule in 2024. This is a unit of energy. A very big unit of energy. Do not worry about the details for now.


We learn more about this later. If the USA has 340 million inhabitants that same year we have
(run the following in your Python environment or Notebook)

In [2]:
100 / (340 * 10**6)

2.941176470588235e-07

Exajoule per person. Since this is a bit cumbersome to read, it is way to small and written in a scientific notation with this e-07 thing, let us convert to Gigajoule (same thing, just smaller unit) for which we multiply by 10\*\*9

In [3]:
100 / (340 * 10**6) * 10**9 # this will come out with way too many decimal places, so we can round it to 2 decimal places

294.1176470588235

And you just learned another very very important fact-of-life about programming: **COMMENT YOUR CODE**.

:::{admonition} What's commenting ? And why is is super important ?
:class: attention

A comment in Python is the annotation after the symbol #, intended to make the code easier for a programmer to understand. This is written into your code cell but not interpreted by the computer as code. It is very important however that others, and even yourself, can understand your own code. You will notice in time, when you come back a week later, to code you have previously written, you will be grateful to your past self for commenting adequately.
:::

### 📖 Interlude on scientific notation and comma use

**On scientific notation**. 

Scientific notation is just a lazy (and smart) way to write massive numbers without having to count a blurry string of zeros.

In Python, the letter e simply stands for "times 10 to the power of."

The Basic Rule: [Number]e[Number of decimal places to move right]

Examples:

1e6 means 1 followed by 6 zeros (1,000,000).

2.5e9 means 2.5 with the decimal moved 9 places to the right (2,500,000,000).

NOTE: This is very confusing as 10 ** 6 (without spaces) is also 1,000,000. So do not confuse 10e6 with one million, this would be ten million!

**On use of commata and full stops** 

When transitioning from French to English, especially in mathematics, economics, and programming, the rules for commas and full stops with numbers (periods) swap. 

1. The Number Rule (Crucial for Coding)
In French, you use a comma for decimals and spaces for thousands. 
In English, you use a **full stop (period)** for decimals and **commas** for thousands.

* **French Format:** `1 000 000,50` *(Un million virgule cinquante)*
* **English Format:** `1,000,000.50` *(One million point fifty)*

Why this matters in Python:
Python uses the English system. If you try to write Pi using a French decimal format (`pi = 3,14`), Python does not see a math number. Instead, the comma tells Python to create a data structure containing two separate numbers: a 3 and a 14. 
* **Wrong (Causes Errors):** `price = 10,50`
* **Correct (Math Decimal):** `price = 10.50`

In [4]:
### for good practice, usually you would want to round to 2 decimal places, so we can use the round() function to do that
### do not worry about the details what a function is for now, we will cover that in the next section, but the syntax of the round() function is round(number, number_of_decimal_places)
round(100 / (340 * 10**6) * 10**9, 2) # this is the same as the previous line, but rounded to 2 decimal places

294.12

### 📖 Putting things together and example exercise

Now getting back to the number we calculated above 294.something. Wow, the average American consumes roughly 300 GJ energy per person! (we rounded up generously). The global average is only around


In [5]:
592/(8.2*10**9) * 10**9 ### this is the total world energy consumption in exajoule (EJ) (592) divided by the total world population of 8.2 billion people, multiplied by 10^9 to convert from EJ to GJ, which is a more manageable unit for per capita energy consumption 

72.19512195121952

~72Gigajoule per person.

First exercise, clarify the above calculation with its respective units. Hint the denominator is global population and the numerator is in a big energy unit called Exajoule. Clarify here means, write down in words what units are in the calculations so for instance if the units were breads baked per kilo of dough, it would be "number of breads / 1 kg of dough". This is supposed help you start right away with thinking quantitatively.

```{dropdown} Solution explained in words

The numerator, `592`, represents the total global primary energy consumption in Exajoules (EJ). The denominator, `8.2 * 10**9`, is the global population.

So, the expression `592 / (8.2 * 10**9)` is calculating `(Total Global EJ) / (Global Population) = EJ per person`.

To convert this result from Exajoules to Gigajoules, you multiply by $10^9$, because $1 \text{ Exajoule} = 10^9 \text{ Gigajoules}$.

So, the full calculation is `(EJ per person) * (Gigajoules per Exajoule)`, which gives you the final unit of `Gigajoules per person`.

```

### 🧩 Exercises

So for the next exercises you conduct simple thought experiments:

1. How much larger would be the total yearly global primary consumption, if everyone lived an American livestyle?
2. How much larger would total yearly $CO_2$ emissions be assuming current average $CO_2$ - intensity of energy (that is the quantity of $CO_2$ emitted per unit of energy - more on this later)?

You may use, or should use, Python as a calculator to help you solve this.

Think about what numbers to multiply or divide and use commenting to clarify what you are doing. A comment line is started with the symbol #. Only open the solutions cell below if you tried yourself for a while or in the best case have a proposed solution.

:::{admonition} Exercises notebook setup
:class: important
From now on for this and all following exercise you should create a Jupyter notebook entitled "exercises_intro" in which you will solve the exercises in the chapter "Introduction to Python and Programming" by yourself. You can also download the Jupyter notebooks in the chapter pressing the download button in the upper right corner. But please try to solve all exercises yourself and no use of AI.
:::



### 📝 Solutions

In [6]:
# Python solution for Exercise 1

# Step 1: Calculate US per capita energy consumption
# 100 EJ divided by 340 million people
100 / 340e6
# Step 2: Calculate New Global Energy Consumption 
# Global population (8.2 billion) multiplied by US per capita consumption
8.2e9 * (100 / 340e6)
# Step 3: Calculate the ratio by dividing the new consumption by the current global consumption (592 EJ)
(8.2e9 * (100 / 340e6)) / 592

# Python demonstration for Exercise 2

# We can prove the math by using a dummy CO2 intensity of 1.0 for both calculations.
# New Emissions divided by Current Emissions:
# [New Energy * 1.0] / [Current Energy * 1.0]
((8.2e9 * (100 / 340e6)) * 1.0) / (592 * 1.0)
# The result is mathematically identical to Exercise 1!

4.073926868044515

```{dropdown} Solutions explained in words

**Exercise 1: Global Energy Consumption with American Lifestyle**

To solve this, we need to calculate the total energy if the entire global population had the US per capita energy consumption.

* **Global Population:** $8.2 \times 10^9$ people
* **US Energy Consumption per capita:** The average American consumes about $300$ GJ of energy per year. This is equivalent to $\frac{100 \text{ EJ}}{340 \times 10^6 \text{ people}}$.

**Calculation:**
New Global Energy Consumption = (Global Population) $\times$ (US Energy Consumption per capita)
New Global Energy Consumption = $(8.2 \times 10^9 \text{ people}) \times (\frac{100 \text{ EJ}}{340 \times 10^6 \text{ people}})$
New Global Energy Consumption $\approx 2412$ EJ

To find out how much larger this is, we compare it to the current global consumption:
Ratio = $\frac{\text{New Global Energy Consumption}}{\text{Current Global Energy Consumption}}$
Ratio = $\frac{2412 \text{ EJ}}{592 \text{ EJ}} \approx 4.07$

So, the total global primary energy consumption would be about **4.1 times larger** if everyone lived an American lifestyle.

**Exercise 2: Global $CO_2$ Emissions with American Lifestyle**

NO CALCULATION NECESSARY! This problem assumes a constant $CO_2$ intensity of energy, which means the amount of $CO_2$ emitted per unit of energy consumed remains the same.

* Current $CO_2$ Emissions = (Current Energy Consumption) $\times$ ($CO_2$ Intensity)
* New $CO_2$ Emissions = (New Energy Consumption) $\times$ ($CO_2$ Intensity)

Since the $CO_2$ intensity is a constant value in this thought experiment, the ratio of emissions will be the same as the ratio of energy consumption.

Ratio of Emissions = $\frac{\text{New } CO_2 \text{ Emissions}}{\text{Current } CO_2 \text{ Emissions}} = \frac{\text{New Energy Consumption}}{\text{Current Energy Consumption}}$

Based on the calculation from Exercise 1, this ratio is approximately **4.1**. Therefore, total global $CO_2$ emissions would also be about **4.1 times larger**.

```